In [1]:
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

# Data cleaning

### Loading the CSV file

In [42]:
path = ""
file_name = "OpenFood.csv"
data_path = path+file_name
df = pd.read_csv(data_path, sep = '\t')

/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (0,3,5,24,25,26,28,36,41,49) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
df.to_pickle("openFood.pkl")

In [4]:
df = pd.read_pickle("openFood.pkl")

### DataFrame Exploration

In [5]:
df.head()

code                                                url   creator  \
0  0000000000017  http://world-en.openfoodfacts.org/product/0000...   kiliweb   
1  0000000000031  http://world-en.openfoodfacts.org/product/0000...  isagoofy   
2  0000000000123  http://world-en.openfoodfacts.org/product/0000...   kiliweb   
3  0000000000178  http://world-en.openfoodfacts.org/product/0000...   kiliweb   
4  0000000000291  http://world-en.openfoodfacts.org/product/0000...   kiliweb   

    created_t      created_datetime last_modified_t last_modified_datetime  \
0  1529059080  2018-06-15T10:38:00Z      1529059204   2018-06-15T10:40:04Z   
1  1539464774  2018-10-13T21:06:14Z      1539464817   2018-10-13T21:06:57Z   
2  1535737982  2018-08-31T17:53:02Z      1535737986   2018-08-31T17:53:06Z   
3  1542456332  2018-11-17T12:05:32Z      1542456333   2018-11-17T12:05:33Z   
4  1534239669  2018-08-14T09:41:09Z      1534239732   2018-08-14T09:42:12Z   

            product_name generic_name quantity      ...        \
0       Vitória crackers          NaN      NaN      ...         
1                  Cacao          NaN    130 g      ...         
2  Sauce Sweety chili 0%          NaN      NaN      ...         
3              Mini coco          NaN      NaN      ...         
4              Mendiants          NaN      NaN      ...         

  carbon-footprint_100g nutrition-score-fr_100g nutrition-score-uk_100g  \
0                   NaN                     NaN                     NaN   
1                   NaN                     NaN                     NaN   
2                   NaN                     NaN                     NaN   
3                   NaN                     NaN                     NaN   
4                   NaN                     NaN                     NaN   

  glycemic-index_100g water-hardness_100g choline_100g phylloquinone_100g  \
0                 NaN                 NaN          NaN                NaN   
1                 NaN                 NaN          NaN                NaN   
2                 NaN                 NaN          NaN                NaN   
3                 NaN                 NaN          NaN                NaN   
4                 NaN                 NaN          NaN                NaN   

  beta-glucan_100g inositol_100g carnitine_100g  
0              NaN           NaN            NaN  
1              NaN           NaN            NaN  
2              NaN           NaN            NaN  
3              NaN           NaN            NaN  
4              NaN           NaN            NaN  

[5 rows x 173 columns]

### DataFrame properties

Are there some duplicated rows?

In [6]:
np.sum(df.duplicated().values == True)

26

So the duplicated values are deleted:

In [7]:
df.drop_duplicates(inplace = True)

We can also drop the column that only conain NaN values:

In [8]:
df.dropna(axis = 1, how='all', inplace = True)

In [9]:
df.shape

(699098, 169)

We now have a total of 699098 rows and 169 columns

Now can the code columnm be set as the index. In order to be so it must be unique:

In [10]:
df.code.is_unique

False

The codes are not unique so let's check what those values are:

In [11]:
duplicate_codes_indices = np.where(df.duplicated('code', keep=False).values)[0]
dups = df.iloc[duplicate_codes_indices].copy()
dups.code = dups.code.astype(float)
dups.sort_values(by = 'code', inplace = True)
dups

code                                                url  \
66475   5.000000e+01  http://world-en.openfoodfacts.org/product/0050...   
535902  5.000000e+01       http://world-en.openfoodfacts.org/product/50   
596588  7.340000e+02      http://world-en.openfoodfacts.org/product/734   
167953  7.340000e+02  http://world-en.openfoodfacts.org/product/0734...   
10873   1.160220e+05  http://world-en.openfoodfacts.org/product/0011...   
209616  1.160220e+05  http://world-en.openfoodfacts.org/product/1160...   
148839  6.356460e+05  http://world-en.openfoodfacts.org/product/0635...   
75132   6.356460e+05  http://world-en.openfoodfacts.org/product/0063...   
75134   6.356840e+05  http://world-en.openfoodfacts.org/product/0063...   
148845  6.356840e+05  http://world-en.openfoodfacts.org/product/0635...   
165192  7.228100e+05  http://world-en.openfoodfacts.org/product/0722...   
89530   7.228100e+05  http://world-en.openfoodfacts.org/product/0072...   
112326  8.379960e+05  http://world-en.openfoodfacts.org/product/0083...   
188264  8.379960e+05  http://world-en.openfoodfacts.org/product/0837...   
205342  8.887070e+05  http://world-en.openfoodfacts.org/product/0888...   
117435  8.887070e+05  http://world-en.openfoodfacts.org/product/0088...   
207921  8.984780e+05  http://world-en.openfoodfacts.org/product/0898...   
118000  8.984780e+05  http://world-en.openfoodfacts.org/product/0089...   
141754  3.318233e+06  http://world-en.openfoodfacts.org/product/0331...   
362674  3.318233e+06  http://world-en.openfoodfacts.org/product/3318...   
520415  4.159932e+06  http://world-en.openfoodfacts.org/product/4159...   
142651  4.159932e+06  http://world-en.openfoodfacts.org/product/0415...   
209301  1.111082e+10  http://world-en.openfoodfacts.org/product/1111...   
4306    1.111082e+10  http://world-en.openfoodfacts.org/product/0011...   
12141   1.356200e+10  http://world-en.openfoodfacts.org/product/0013...   
210444  1.356200e+10  http://world-en.openfoodfacts.org/product/1356...   
210445  1.356200e+10  http://world-en.openfoodfacts.org/product/1356...   
12142   1.356200e+10  http://world-en.openfoodfacts.org/product/0013...   
12535   1.410007e+10  http://world-en.openfoodfacts.org/product/0014...   
210625  1.410007e+10  http://world-en.openfoodfacts.org/product/1410...   
...              ...                                                ...   
265815           NaN                                                NaN   
265826           NaN                                                NaN   
265850           NaN                                                NaN   
265863           NaN                                                NaN   
265871           NaN                                                NaN   
265873           NaN                                                NaN   
265879           NaN                                                NaN   
265926           NaN                                                NaN   
265946           NaN                                                NaN   
265956           NaN                                                NaN   
265958           NaN                                                NaN   
265962           NaN                                                NaN   
265964           NaN                                                NaN   
265975           NaN                                                NaN   
265977           NaN                                                NaN   
265984           NaN                                                NaN   
265987           NaN                                                NaN   
266023           NaN                                                NaN   
266065           NaN                                                NaN   
266068           NaN                                                NaN   
266098           NaN                                                NaN   
266100           NaN                     

As it can be seen there are two types of duplicates:
Some codes appear twice and the some code are NaN. As it can be observed when code are duplicated they are often the same and sometimes one row contain more information than the other. So we could just keep the rows that contain the most information. And for the rows with NaN as code, the values of the columns are not matching the column so there must be a shift of values. As a lot of values are missing, it would be complex to recover the true data and there are only tens of code with NaNs, we can delete them:

In [21]:
dups = dups[dups != float('nan')]

dups['nb_of_features'] = np.sum(dups.notnull(), axis=1)
dups.reset_index(inplace = True)
indices_to_remove = dups.sort_values(['code','nb_of_features'],ascending = True)\
                        .groupby('code',as_index=False).first()['index'].values
df.dropna(subset=['code'],inplace=True)
df.drop(indices_to_remove,inplace = True)

In [22]:
df.code.is_unique

True

In [25]:
df.set_index('code',inplace = True)
df.head()

url   creator  \
code                                                                         
0000000000017  http://world-en.openfoodfacts.org/product/0000...   kiliweb   
0000000000031  http://world-en.openfoodfacts.org/product/0000...  isagoofy   
0000000000123  http://world-en.openfoodfacts.org/product/0000...   kiliweb   
0000000000291  http://world-en.openfoodfacts.org/product/0000...   kiliweb   
0000000000949  http://world-en.openfoodfacts.org/product/0000...   kiliweb   

                created_t      created_datetime last_modified_t  \
code                                                              
0000000000017  1529059080  2018-06-15T10:38:00Z      1529059204   
0000000000031  1539464774  2018-10-13T21:06:14Z      1539464817   
0000000000123  1535737982  2018-08-31T17:53:02Z      1535737986   
0000000000291  1534239669  2018-08-14T09:41:09Z      1534239732   
0000000000949  1523440813  2018-04-11T10:00:13Z      1523440823   

              last_modified_datetime               product_name generic_name  \
code                                                                           
0000000000017   2018-06-15T10:40:04Z           Vitória crackers          NaN   
0000000000031   2018-10-13T21:06:57Z                      Cacao          NaN   
0000000000123   2018-08-31T17:53:06Z      Sauce Sweety chili 0%          NaN   
0000000000291   2018-08-14T09:42:12Z                  Mendiants          NaN   
0000000000949   2018-04-11T10:00:23Z  Salade de carottes râpées          NaN   

              quantity packaging      ...       carbon-footprint_100g  \
code                                  ...                               
0000000000017      NaN       NaN      ...                         NaN   
0000000000031    130 g       NaN      ...                         NaN   
0000000000123      NaN       NaN      ...                         NaN   
0000000000291      NaN       NaN      ...                         NaN   
0000000000949      NaN       NaN      ...                         NaN   

              nutrition-score-fr_100g nutrition-score-uk_100g  \
code                                                            
0000000000017                     NaN                     NaN   
0000000000031                     NaN                     NaN   
0000000000123                     NaN                     NaN   
0000000000291                     NaN                     NaN   
0000000000949                     NaN                     NaN   

              glycemic-index_100g water-hardness_100g choline_100g  \
code                                                                 
0000000000017                 NaN                 NaN          NaN   
0000000000031                 NaN                 NaN          NaN   
0000000000123                 NaN                 NaN          NaN   
0000000000291                 NaN                 NaN          NaN   
0000000000949                 NaN                 NaN          NaN   

              phylloquinone_100g beta-glucan_100g inositol_100g carnitine_100g  
code                                                                            
0000000000017                NaN              NaN           NaN            NaN  
0000000000031                NaN              NaN           NaN            NaN  
0000000000123                NaN              NaN           NaN            NaN  
0000000000291                NaN              NaN           NaN            NaN  
0000000000949                NaN              NaN           NaN            NaN  

[5 rows x 168 columns]

### Removing useless columns

In order to answer our research questions, we do not need all the columns such as *creator* or *created_datetime* among others. We decided to drop all the columns that are not relevant to our research questions. This is done manually by looking at the columns in the dataframe.

In [227]:
columns_to_drop = ['url',
                  'creator',
                  'created_t',
                  'created_datetime',
                  'last_modified_t',
                  'last_modified_datetime',
                  'generic_name',
                  'quantity',
                  'countries_tags',
                  'brands_tags',
                  'origins',
                  'countries',
                  'origins_tags',
                  'emb_codes',
                  'emb_codes_tags',
                  'first_packaging_code_geo',
                  'cities',
                  'cities_tags',
                  'purchase_places',
                  'stores',
                  'traces',
                  'traces_tags',
                  'serving_size',
                  'no_nutriments',
                  'additives',
                  'additives_tags',
                  'image_url',
                  'allergens', 
                  'image_small_url',
                  'ph_100g',
                  'manufacturing_places_tags',
                  'pnns_groups_1',
                  'pnns_groups_2',
                  'image_ingredients_url',
                  'image_ingredients_small_url',
                  'image_nutrition_url',
                  'image_nutrition_small_url',
                  'states',        
                  'states_en',
                  'states_tags',
                  'serving_quantity',
                  'traces_en',
                  'allergens_en',
                  'categories',
                  'categories_tags',
                  'packaging_tags',
                  'additives_en',
                  'ingredients_that_may_be_from_palm_oil',
                  'ingredients_from_palm_oil'
                  ]

Some columns that are deleted are obvious not part of our plans (for instance *image_nutrition_url* or *first_packaging_code_geo*) while others required more in-depth analysis. Indeed, a large number of columns are divided into 3 columns with redundancy. For example, the *countries* column also has *countries_en* for the english name and *countries_tags* for the tag representing that country. By analyzing the three columns, we found that *countries_en* has the clearest names (while others have different names for the same country). Hence, we decided to only keep the *countries_en* column. The same analysis has been done for all the columns that present the same situation. Finally, a few columns such as *ingredients_from_palm_oil* only have NaN values and are thus removed from the dataframe.

In [246]:
# We drop all the columns defined above
df_new = df.drop(columns_to_drop, axis=1)
df_new

code                                      product_name  \
0       0000000000017                                  Vitória crackers   
1       0000000000031                                             Cacao   
2       0000000000123                             Sauce Sweety chili 0%   
3       0000000000178                                         Mini coco   
4       0000000000291                                         Mendiants   
5       0000000000949                         Salade de carottes râpées   
6       0000000000970                       Fromage blanc aux myrtilles   
7       0000000001001                                               NaN   
8       0000000001007                                          Vainilla   
9       0000000001137                                 Baguette parisien   
10      0000000001151                     &quot;Baguette Lyonnais&quot;   
11      0000000001199                            Solène céréales poulet   
12      0000000001281                                Tarte noix de coco   
13      0000000001311                        Salade de fruits exotiques   
14      0000000001328                                  Chouquettes x 30   
15      0000000001564                               Fromage blanc pêche   
16      0000000001663                            Crème dessert chocolat   
17      0000000001670                              Bouillie de sarrasin   
18      0000000001885                                  Compote de poire   
19      0000000002042                                  Paëlla de poulet   
20      0000000002103                            Aiguillettes de poulet   
21      0000000002219                                Salade shaker chef   
22      0000000002257                    Salade de macedoine de légumes   
23      0000000002264                                 Baguette Poitevin   
24      0000000002363                                      Suedois thon   
25      0000000002400                                   Ciabatta Bombay   
26      0000000002417                                     Ciabatta Roma   
27      0000000003018                                     Salade tomate   
28      0000000003087                                Farine de blé noir   
29      0000000003384                           Mousse chocolat douceur   
...               ...                                               ...   
699094  9984213030003                                Alwani Sagai Dates   
699095  9984333999679                       Pate feuilletée sans gluten   
699096  9985023697844                                 Barquettes Fraise   
699097  9985043117995  Pâte à tartiner aux noisettes et au cacao maigre   
699098  9990000024742                              Pur cacao biologique   
699099  9990000039692                                 Rooïbos aromatise   
699100       99903336                                   Jus multifruits   
699101  9991111111154                       Compote à Boire Pomme Poire   
699102  9991111114117                                  Nectar d'abricot   
699103  9991111116111                          Puree de pomme framboise   
699104  9996015650648                          Quick milk magic slipper   
699105  9996980313319                                       cup noodles   
699106  9998070000000                                    Miel de fleurs   
699107  9999001900031                                          Formula1   
699108  9999091865142                         Paprikás Kukorica csemege   
699109       99991111                   Trio de lentilles & saumon fumé   
699110  9999200847366                                      Choco pearls   
699111       99993306                       Pizza thon cuite sur pierre   
699112       99996666          Class A'peel Dot Sparkler Stickers-black   
699113  9999729004035                                       Comte vieux   
699114  9999900002553                       Chocolat de Couverture Noir   
699115   999990026839                   Su

This current data-set still has 124 columns, all of them that can be used for our research questions. Currently, the columns are dominated by the ingredients. However, there is a possibility that some ingredients have mostly *NaN* values, thus would not help our analysis and made the data not clean. As a result, we decided to analyze the data for each ingredient i.e. the number of values that are not *NaN* for every ingredients.

### Analyzing the ingredients

Most of the columns at the end represent ingredients that are important for our research questions. However, it seems that most of the ingredients contain too much *NaN* values. We would like to check, for each ingredient, how many non-NaN values it has.

In [235]:
# We take the column indices that are ingredients
l1 = list(range(20, 114))
l2 = list(range(118, df_new.shape[1]))

# We take the columns that represent ingredients to analyze them
df_ingredients = df_new[df_new.columns[l1 + l2]]

# For each ingredient, we see how much times it appears (how many non-NaN values)
ingredients_count = df_ingredients.count().sort_values(ascending=False)

# We see the 30 ingredients that have the less NaN values
ingredients_count[0:30]

energy_100g                             596006
proteins_100g                           594569
fat_100g                                590631
carbohydrates_100g                      590377
sugars_100g                             579971
salt_100g                               574841
sodium_100g                             574806
saturated-fat_100g                      564706
fiber_100g                              274143
cholesterol_100g                        144627
trans-fat_100g                          143944
calcium_100g                            143501
vitamin-c_100g                          141726
iron_100g                               141616
vitamin-a_100g                          138331
potassium_100g                           25420
polyunsaturated-fat_100g                 23788
monounsaturated-fat_100g                 23749
vitamin-pp_100g                          12299
vitamin-b1_100g                          11841
vitamin-b2_100g                          11401
alcohol_100g 

We notice that some ingredients only have values for 1 product out of 699098. In fact, most of the ingredients do not have values for more than 4000 products out of 699098 in total. These columns can then be dropped because they are not interesting values since they do not add too much value. Taking the top 30 was arbitrarily chosen because 30 is around 1/3 of the total number of ingredients which is good enough for our analysis

In [244]:
# We convert the above series into a dataframe to easily take the names of the ingredients and put them
# in a list. Note that we take the ingredients that have less than 4000 non-NaN values.
ingredients_count_df = pd.DataFrame(ingredients_count).reset_index()[30:94]

# We take the ingredients that are not in top 30 and put them in a list
ingredients_to_drop = ingredients_count_df['index'].tolist()

# We drop all those ingredients
df_no_ingredients = df_new.drop(ingredients_to_drop, axis=1)
df_no_ingredients.head(10)

code                 product_name packaging brands categories_en  \
0  0000000000017             Vitória crackers       NaN    NaN           NaN   
1  0000000000031                        Cacao       NaN    NaN           NaN   
2  0000000000123        Sauce Sweety chili 0%       NaN    NaN           NaN   
3  0000000000178                    Mini coco       NaN    NaN           NaN   
4  0000000000291                    Mendiants       NaN    NaN           NaN   
5  0000000000949    Salade de carottes râpées       NaN    NaN           NaN   
6  0000000000970  Fromage blanc aux myrtilles       NaN    NaN           NaN   
7  0000000001001                          NaN       NaN    NaN           NaN   
8  0000000001007                     Vainilla       NaN    NaN           NaN   
9  0000000001137            Baguette parisien       NaN    NaN           NaN   

  manufacturing_places labels labels_tags labels_en countries_en  \
0                  NaN    NaN         NaN       NaN       France   
1                  NaN    NaN         NaN       NaN       France   
2                  NaN    NaN         NaN       NaN       France   
3                  NaN    NaN         NaN       NaN       France   
4                  NaN    NaN         NaN       NaN       France   
5                  NaN    NaN         NaN       NaN       France   
6                  NaN    NaN         NaN       NaN       France   
7                  NaN    NaN         NaN       NaN       France   
8                  NaN    NaN         NaN       NaN       France   
9                  NaN    NaN         NaN       NaN       France   

  ingredients_text  additives_n  ingredients_from_palm_oil_n  \
0              NaN          NaN                          NaN   
1              NaN          NaN                          NaN   
2              NaN          NaN                          NaN   
3              NaN          NaN                          NaN   
4              NaN          NaN                          NaN   
5              NaN          NaN                          NaN   
6              NaN          NaN                          NaN   
7              NaN          NaN                          NaN   
8              NaN          NaN                          NaN   
9              NaN          NaN                          NaN   

  ingredients_from_palm_oil_tags  ingredients_that_may_be_from_palm_oil_n  \
0                            NaN                                      NaN   
1                            NaN                                      NaN   
2                            NaN                                      NaN   
3                            NaN                                      NaN   
4                            NaN                                      NaN   
5                            NaN                                      NaN   
6                            NaN                                      NaN   
7                            NaN                                      NaN   
8                            NaN                                      NaN   
9                            NaN                                      NaN   

  ingredients_that_may_be_from_palm_oil_tags  nutrition_grade_uk  \
0                                        NaN                 NaN   
1                                        NaN                 NaN   
2                                        NaN                 NaN   
3                                        NaN                 NaN   
4                                        NaN                 NaN   
5                                        NaN                 NaN   
6                                        NaN                 NaN   
7                                        NaN                 NaN   
8                                        NaN                 NaN   
9                                        NaN                 NaN   

  nutrition_grade_fr main_category main_category_en  energy_100g  fat_100g  \
0 

This new dataframe has 60 columns, which is an improvement form the original 169 columns. More than 100 columns were either useless for our research questions or do not have enough data to be interesting.